File: elixir/day3/lib/part2.ex
defmodule Part2 do
  def extract_and_sum_multiplications(input) do
    # First find all instructions in order
    instructions =
      Regex.scan(~r/(?:do\(\)|don't\(\)|mul\(\d+,\d+\))/, input)
      |> List.flatten()

    # Process each instruction in order
    {_enabled, sum} =
      Enum.reduce(instructions, {true, 0}, fn
        "do()", {_enabled, sum} ->
          {true, sum}

        "don't()", {_enabled, sum} ->
          {false, sum}

        "mul(" <> rest, {enabled, sum} ->
          if enabled do
            case Regex.run(~r/(\d+),(\d+)\)/, rest, capture: :all_but_first) do
              [x, y] -> {enabled, sum + String.to_integer(x) * String.to_integer(y)}
              _ -> {enabled, sum}
            end
          else
            {enabled, sum}
          end

        _, acc ->
          acc
      end)

    sum
  end

  def part2(input) do
    extract_and_sum_multiplications(input)
  end
end

File.read!(~c"input.txt")
|> Part2.part2()
|> IO.inspect(label: "part2")

File: ruby/day3/part2.rb
# frozen_string_literal: true

def part2(input)
  input.scan(/(do\(\)|don't\(\)|mul\((\d{1,3}),(\d{1,3})\))/).reduce([true, 0]) do |(enabled, sum), command|
    case command[0]
    when 'do()'
      [true, sum]
    when "don't()"
      [false, sum]
    else
      [enabled, sum + (enabled ? command[1].to_i * command[2].to_i : 0)]
    end
  end.last
end

if __FILE__ == $PROGRAM_NAME
  input = File.read('input.txt')
  puts "Result: #{part2(input)}"
end
